src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,351 @@
+                        AssemblyName.GetAssemblyName(dll);
+                    }
+                    catch
+                    {
+                        // The file is not a valid .NET assembly, skip it.
+                        continue;
+                    }
+
+                    analyzerAssemblyLoader.AddDependencyLocation(dll);
+                }
+
+                extension = new CustomMessageHandlerExtension(analyzerAssemblyLoader);
+                _extensions[assemblyFolderPath] = extension;
+            }
+        }
+
+        // AssemblyLoadLockObject is only used to avoid multiple calls from the same extensions to load the same assembly concurrently
+        // resulting in the constructors of the same handlers being called more than once.
+        // All other concurrent operations, including modifying extension.Assemblies are protected by _lockObject.
+        lock (extension.AssemblyLoadLockObject)
+        {
+            if (extension.Assemblies.TryGetValue(assemblyFileName, out var extensionAssembly))
+            {
+                if (extensionAssembly.HasValue)
+                {
+                    return ValueTask.FromResult<RegisterExtensionResponse>(
+                        new(extensionAssembly.Value.WorkspaceMessageHandlers.ToImmutableArray(), extensionAssembly.Value.DocumentMessageHandlers.ToImmutableArray()));
+                }
+                else
+                {
+                    throw new InvalidOperationException($"A previous attempt to load {assemblyFilePath} failed.");
+                }
+            }
+            else
+            {
+                var mustCleanupExtension = false;
+                try
+                {
+                    var assembly = extension.AnalyzerAssemblyLoader.LoadFromPath(assemblyFilePath);
+                    var messageHandlers = _customMessageHandlerFactory.CreateWorkspaceMessageHandlers(assembly)
+                        .ToDictionary(h => h.Name, h => h);
+                    var messageDocumentHandlers = _customMessageHandlerFactory.CreateDocumentMessageHandlers(assembly)
+                        .ToDictionary(h => h.Name, h => h);
+
+                    // Important, you can lock _lockObject when holding a lock on AssemblyLoadLockObject, not vice-versa
+                    lock (_lockObject)
+                    {
+                        // Make sure a call to UnloadCustomMessageHandlersAsync hasn't happened while we relinquished the lock on _lockObject
+                        if (!_extensions.TryGetValue(assemblyFolderPath, out var currentExtension) || !currentExtension.Equals(extension))
+                        {
+                            // extension is not in the _extensions dictionary anymore, so it's AnalyzerAssemblyLoader must be unloaded
+                            mustCleanupExtension = true;
+                            throw new InvalidOperationException($"{assemblyFilePath} was unloaded while loading handlers.");
```

Comment:
ok.  this feels very strange.  can we just do everything we need under a particular lock so that all transitions are atomic?

basically, it seems like it would be nice to just have an outer lock that just gives you th extension instance.  you then ask *it* to do what it needs, and it has its state atomatic inside of itself.  If it then gets tossed from teh dict, that todesn't impact existing instances of the extension type wrapper.